fairroot/tools/FairMemory.h

```diff
@@ -0,0 +1,67 @@
+/********************************************************************************
+ *    Copyright (C) 2023 GSI Helmholtzzentrum fuer Schwerionenforschung GmbH    *
+ *                                                                              *
+ *              This software is distributed under the terms of the             *
+ *              GNU Lesser General Public Licence (LGPL) version 3,             *
+ *                  copied verbatim in the file "LICENSE"                       *
+ ********************************************************************************/
+
+#ifndef FAIRROOT_MEMORY_H_
+#define FAIRROOT_MEMORY_H_
+
+#include <memory>
+#include <optional>
+
+/**
+ * \brief Internal implementation details
+ *
+ * Any declarations in here are strictly for internal usage.
+ * They may never enter public APIs.
+ *
+ * \note Anything in here can change at any time without notice!
+ */
+namespace fairroot::detail
+{
+
+template<typename T>
+struct optional_deleter
+{
+    std::optional<std::default_delete<T>> deleter{std::in_place};
+
+    optional_deleter() = default;
+    optional_deleter(const std::default_delete<T>& x)
+        : deleter(std::in_place, x)
+    {}
+    optional_deleter(std::nullopt_t)
+        : deleter()
+    {}
+
+    void operator()(T* ptr) const
+    {
+        if (deleter) {
+            (*deleter)(ptr);
+        }
+    }
+};
```

Comment:
```suggestion
template <typename T, typename Deleter>
struct optional_deleter
{
    std::optional<Deleter> deleter{std::in_place};

    optional_deleter() = default;
    optional_deleter(const Deleter& x)
        : deleter(std::in_place, x)
    {}
    optional_deleter(std::nullopt_t)
        : deleter()
    {}

    void operator()(T* ptr) const
    {
        if (deleter) {
            (*deleter)(ptr);
        }
    }
};
```

* Or is there a good reason to hard-code `std::default_delete` here I am missing+?
* Perhaps there is even some way to inspect the `T` from `Deleter`? (to just have `template <typename Deleter> struct optional_deleter`)

Reply:
> * Or is there a good reason to hard-code std::default_delete here which I am missing?

Only KISS. We hopefully wont ever need optional other deleters in FairRoot.

As I (intentionally) named this thing more like in the `std::` namespace (and not using the FairRoot conventions), I wrote it with the hidden idea that others could copy and re-use this code, if they needed that functionality. So being more generic is good!

> * Perhaps there is even some way to inspect the T from Deleter? 

```diff
-    void operator()(T* ptr) const
+    void operator()(auto* ptr) const
```

this would remove the last trace of `T` from `optional_deleter`. But this it seems, that this is only possible with C++20. So we can't have this yet.

So for I changed it to
```c++
template<typename T, typename Deleter = std::default_delete<T>>
```

so that it can be used without the deleter if wanted.